File: android/app/src/main/java/com/mochi/keyboard/features/community/CommunityViewModel.kt
package com.mochi.keyboard.features.community

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mochi.keyboard.data.AuthRepository
import com.mochi.keyboard.data.FollowRepository
import com.mochi.keyboard.data.LikeRepository
import com.mochi.keyboard.data.ReportRepository
import com.mochi.keyboard.data.ThemeCache
import com.mochi.keyboard.data.ThemeRepository
import com.mochi.keyboard.model.CommunityPost
import com.mochi.keyboard.model.KeyboardTheme
import com.mochi.keyboard.model.TagPalette
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlin.math.abs

sealed interface CommunityUiState {
    data object Loading : CommunityUiState
    data class Data(
        val feedThemes: List<KeyboardTheme>,
        val latestThemes: List<KeyboardTheme>,
        val creators: List<CommunityCreatorUi>
    ) : CommunityUiState
    data class Error(val message: String) : CommunityUiState
}

/** No `isVerified`/avatar-photo field exists anywhere in this app's schema, so real creator tiles
 * never show a verified badge or a real photo (falls back to the same neutral avatar box every
 * unmatched CreatorAvatar id already renders) - not an oversight, there's nothing to source it from. */
data class CommunityCreatorUi(val uid: String, val name: String, val themeCount: Int, val isFollowing: Boolean)

/**
 * Backs Community's 5 feed tabs + Popular Creators + Latest Creations + follow/report actions.
 * ThemeRepository's getTopRanked/getPublishedThemes/getThemesByCreators/getThemesByIds map 1:1 onto
 * Popular/Latest/Following/My Likes (their doc comments were written anticipating this exact screen
 * in WA2). "For you" has no dedicated query of its own - stands in with getTopRanked until a real
 * personalized feed is in scope.
 *
 * Popular Creators has no dedicated backend query either (UserRepository can only create a profile,
 * not list/search them) - derived instead from the top-ranked themes batch, grouping by creatorUid
 * and counting appearances within that batch. That's a real-data approximation (not MockData), just
 * bounded to whichever creators show up in the top 20 themes, not a true global ranking.
 */
class CommunityViewModel(
    private val themeRepository: ThemeRepository,
    private val likeRepository: LikeRepository,
    private val followRepository: FollowRepository,
    private val reportRepository: ReportRepository,
    private val authRepository: AuthRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow<CommunityUiState>(CommunityUiState.Loading)
    val uiState: StateFlow<CommunityUiState> = _uiState.asStateFlow()

    private var followedUids: Set<String> = emptySet()
    private var selectedTab: String = "For you"

    init {
        viewModelScope.launch {
            authRepository.currentUser?.uid?.let { uid ->
                runCatching { followRepository.followedUids(uid) }.onSuccess { followedUids = it.toSet() }
            }
            loadFeed()
        }
    }

    fun selectTab(tab: String) {
        if (tab == selectedTab) return
        selectedTab = tab
        loadFeed()
    }

    fun loadFeed() {
        viewModelScope.launch {
            val uid = authRepository.currentUser?.uid
            val feedResult = runCatching { fetchTab(selectedTab, uid) }
            val feedThemes = feedResult.getOrNull()
            if (feedThemes == null) {
                _uiState.value = CommunityUiState.Error(feedResult.exceptionOrNull()?.message ?: "Couldn't load community feed.")
                return@launch
            }
            val topRanked = runCatching { themeRepository.getTopRanked(20) }.getOrDefault(emptyList())
            val latest = runCatching { themeRepository.getPublishedThemes(10) }.getOrDefault(emptyList())
            ThemeCache.put(feedThemes)
            ThemeCache.put(topRanked)
            ThemeCache.put(latest)
            _uiState.value = CommunityUiState.Data(
                feedThemes = feedThemes,
                latestThemes = latest,
                creators = buildCreators(topRanked)
            )
        }
    }

    private suspend fun fetchTab(tab: String, uid: String?): List<KeyboardTheme> = when (tab) {
        "Popular" -> themeRepository.getTopRanked(20)
        "Latest" -> themeRepository.getPublishedThemes(20)
        "Following" -> themeRepository.getThemesByCreators(if (uid != null) followRepository.followedUids(uid) else emptyList())
        "My Likes" -> themeRepository.getThemesByIds(if (uid != null) likeRepository.likedThemeIds(uid) else emptyList())
        else -> themeRepository.getTopRanked(20) // "For you" - no dedicated query yet, see class doc
    }

    private fun buildCreators(themes: List<KeyboardTheme>): List<CommunityCreatorUi> =
        themes.filter { it.creatorUid.isNotBlank() }
            .groupBy { it.creatorUid }
            .map { (uid, group) ->
                CommunityCreatorUi(uid = uid, name = group.first().creatorName, themeCount = group.size, isFollowing = followedUids.contains(uid))
            }

    fun toggleFollow(creatorUid: String) {
        val uid = authRepository.currentUser?.uid ?: return
        if (uid == creatorUid) return
        val wasFollowing = followedUids.contains(creatorUid)
        followedUids = if (wasFollowing) followedUids - creatorUid else followedUids + creatorUid
        reapplyFollowState()
        viewModelScope.launch {
            runCatching { if (wasFollowing) followRepository.unfollow(uid, creatorUid) else followRepository.follow(uid, creatorUid) }
                .onFailure {
                    followedUids = if (wasFollowing) followedUids + creatorUid else followedUids - creatorUid
                    reapplyFollowState()
                }
        }
    }

    private fun reapplyFollowState() {
        val current = _uiState.value
        if (current is CommunityUiState.Data) {
            _uiState.value = current.copy(creators = current.creators.map { it.copy(isFollowing = followedUids.contains(it.uid)) })
        }
    }

    fun reportTheme(themeId: String, reason: String) {
        val uid = authRepository.currentUser?.uid ?: return
        viewModelScope.launch { runCatching { reportRepository.reportTheme(uid, themeId, reason) } }
    }
}

/** Community's card models carry a bit more than a bare theme (a fixed hashtag-chip color, a
 * summary line) that ThemeDocument has no direct field for - description doubles as the summary,
 * and the tag palette is deterministically derived from the theme id so a given theme always shows
 * the same chip color rather than flickering between reloads. */
internal fun KeyboardTheme.toCommunityPost(): CommunityPost = CommunityPost(
    id = id,
    name = name,
    creatorName = creatorName,
    thumbAssetName = imageAssetName,
    summary = description,
    likeCount = likeCount,
    hashtags = hashtags,
    tagPalette = TagPalette.entries[abs(id.hashCode()) % TagPalette.entries.size]
)

File: ios/MochiApp/Features/Community/CommunityViewModel.swift
import Foundation

/// No `isVerified`/avatar-photo field exists anywhere in this app's schema, so real creator tiles
/// never show a verified badge or a real photo (falls back to the same neutral avatar every
/// unmatched creator id already renders) — not an oversight, there's nothing to source it from.
struct CommunityCreatorUi: Identifiable {
    let uid: String
    let name: String
    let themeCount: Int
    let isFollowing: Bool
    var id: String { uid }
}

enum CommunityUiState {
    case loading
    case data(feedThemes: [KeyboardTheme], latestThemes: [KeyboardTheme], creators: [CommunityCreatorUi])
    case error(String)
}

/// Swift mirror of android/.../features/community/CommunityViewModel.kt — backs Community's 5 feed
/// tabs + Popular Creators + Latest Creations + follow/report actions.
///
/// ThemeRepository's getTopRanked/getPublishedThemes/getThemesByCreators/getThemesByIds map 1:1 onto
/// Popular/Latest/Following/My Likes. "For you" has no dedicated query of its own — stands in with
/// getTopRanked until a real personalized feed is in scope.
///
/// Popular Creators has no dedicated backend query either (UserRepository can only create a profile,
/// not list/search them) — derived instead from the top-ranked themes batch, grouping by creatorUid
/// and counting appearances within that batch. That's a real-data approximation (not MockData), just
/// bounded to whichever creators show up in the top 20 themes, not a true global ranking.
@MainActor
final class CommunityViewModel: ObservableObject {
    @Published private(set) var uiState: CommunityUiState = .loading

    private let themeRepository: ThemeRepository?
    private let likeRepository: LikeRepository?
    private let followRepository: FollowRepository?
    private let reportRepository: ReportRepository?
    private let authRepository: AuthRepository?

    private var followedUids: Set<String> = []
    private var selectedTab: String = "For you"

    init(container: AppContainer?) {
        themeRepository = container?.themeRepository
        likeRepository = container?.likeRepository
        followRepository = container?.followRepository
        reportRepository = container?.reportRepository
        authRepository = container?.authRepository

        guard themeRepository != nil else { return }
        Task {
            if let uid = authRepository?.currentUser?.uid, let followRepository {
                let uids = (try? await followRepository.followedUids(followerId: uid)) ?? []
                followedUids = Set(uids)
            }
            await loadFeed()
        }
    }

    func selectTab(_ tab: String) {
        guard tab != selectedTab else { return }
        selectedTab = tab
        Task { await loadFeed() }
    }

    func loadFeed() async {
        guard let themeRepository else { return }
        let uid = authRepository?.currentUser?.uid
        do {
            let feedThemes = try await fetchTab(selectedTab, uid: uid)
            async let topRankedTask = try? themeRepository.getTopRanked(limit: 20)
            async let latestTask = try? themeRepository.getPublishedThemes(limit: 10)
            let topRanked = await topRankedTask ?? []
            let latest = await latestTask ?? []
            uiState = .data(feedThemes: feedThemes, latestThemes: latest, creators: buildCreators(topRanked))
        } catch {
            uiState = .error(error.localizedDescription)
        }
    }

    private func fetchTab(_ tab: String, uid: String?) async throws -> [KeyboardTheme] {
        guard let themeRepository else { return [] }
        switch tab {
        case "Popular":
            return try await themeRepository.getTopRanked(limit: 20)
        case "Latest":
            return try await themeRepository.getPublishedThemes(limit: 20)
        case "Following":
            var uids: [String] = []
            if let uid, let followRepository {
                uids = (try? await followRepository.followedUids(followerId: uid)) ?? []
            }
            return try await themeRepository.getThemesByCreators(creatorUids: uids)
        case "My Likes":
            var ids: [String] = []
            if let uid, let likeRepository {
                ids = (try? await likeRepository.likedThemeIds(uid: uid)) ?? []
            }
            return try await themeRepository.getThemesByIds(ids)
        default: // "For you" — no dedicated query yet, see class doc
            return try await themeRepository.getTopRanked(limit: 20)
        }
    }

    private func buildCreators(_ themes: [KeyboardTheme]) -> [CommunityCreatorUi] {
        var order: [String] = []
        var counts: [String: Int] = [:]
        var names: [String: String] = [:]
        for theme in themes where !theme.creatorUid.isEmpty {
            if counts[theme.creatorUid] == nil {
                order.append(theme.creatorUid)
                names[theme.creatorUid] = theme.creatorName
            }
            counts[theme.creatorUid, default: 0] += 1
        }
        return order.map { uid in
            CommunityCreatorUi(uid: uid, name: names[uid] ?? "", themeCount: counts[uid] ?? 0, isFollowing: followedUids.contains(uid))
        }
    }

    func toggleFollow(_ creatorUid: String) {
        guard let uid = authRepository?.currentUser?.uid, uid != creatorUid, let followRepository else { return }
        let wasFollowing = followedUids.contains(creatorUid)
        if wasFollowing { followedUids.remove(creatorUid) } else { followedUids.insert(creatorUid) }
        reapplyFollowState()
        Task {
            do {
                if wasFollowing {
                    try await followRepository.unfollow(followerId: uid, followeeId: creatorUid)
                } else {
                    try await followRepository.follow(followerId: uid, followeeId: creatorUid)
                }
            } catch {
                if wasFollowing { followedUids.insert(creatorUid) } else { followedUids.remove(creatorUid) }
                reapplyFollowState()
            }
        }
    }

    private func reapplyFollowState() {
        guard case .data(let feedThemes, let latestThemes, let creators) = uiState else { return }
        let updated = creators.map { CommunityCreatorUi(uid: $0.uid, name: $0.name, themeCount: $0.themeCount, isFollowing: followedUids.contains($0.uid)) }
        uiState = .data(feedThemes: feedThemes, latestThemes: latestThemes, creators: updated)
    }

    /// Mirrors a like to Firestore. The local record (`LikedThemeStore`) is written by the view
    /// first and is what the heart reads, so this is a no-op whenever there's no backend or no
    /// signed-in user rather than a path the UI has to wait on.
    func setLiked(themeId: String, liked: Bool) {
        guard let uid = authRepository?.currentUser?.uid, let likeRepository else { return }
        Task {
            if liked {
                try? await likeRepository.like(uid: uid, themeId: themeId)
            } else {
                try? await likeRepository.unlike(uid: uid, themeId: themeId)
            }
        }
    }

    func reportTheme(themeId: String, reason: String) {
        guard let uid = authRepository?.currentUser?.uid, let reportRepository else { return }
        Task { try? await reportRepository.reportTheme(reporterUid: uid, themeId: themeId, reason: reason) }
    }
}

/// Community's card model carries a summary line that `KeyboardTheme` has no direct field for —
/// the theme's description doubles as it. Hashtag chip colors are positional and applied at render
/// time (see `CommunityPost`), so nothing colour-related is derived here.
extension KeyboardTheme {
    func toCommunityPost() -> CommunityPost {
        CommunityPost(
            id: id,
            name: name,
            creatorName: creatorName,
            thumbAssetName: imageAssetName,
            summary: description,
            likeCount: likeCount,
            hashtags: hashtags
        )
    }
}
